File: playwright-main/stealth-kit/src/behavior/scroll.ts
import type { Page } from 'playwright-core';

/**
 * Human-like scrolling: several wheel ticks with varied deltas and pauses
 * instead of one instant jump to a target offset.
 */

const sleep = (ms: number) => new Promise(r => setTimeout(r, ms));
const rand = (min: number, max: number) => min + Math.random() * (max - min);

export interface HumanScrollOptions {
  /** Total vertical distance in px. Default: one viewport-ish, downward. */
  distance?: number;
  direction?: 'down' | 'up';
  /** Split the distance into at most this many ticks. */
  maxTicks?: number;
}

export async function humanScroll(page: Page, options: HumanScrollOptions = {}): Promise<void> {
  const sign = options.direction === 'up' ? -1 : 1;
  let remaining = Math.abs(options.distance ?? rand(400, 900));
  const maxTicks = options.maxTicks ?? 8;
  let ticks = 0;

  while (remaining > 0 && ticks < maxTicks) {
    // Early ticks are larger; the last ones fine-tune, like a real reader.
    const delta = ticks === 0 ? rand(0.4, 0.6) * remaining : rand(0.15, 0.45) * remaining;
    await page.mouse.wheel(0, sign * Math.max(40, delta));
    remaining -= delta;
    ticks++;
    await sleep(rand(120, 450));
  }
}

File: Agent-Reach-main/stealth_kit/behavior/scroll.py
"""
Human-like scrolling: several wheel ticks with varied deltas and pauses
instead of one instant jump to a target offset.

Port of stealth-kit/src/behavior/scroll.ts. Randomness comes from the
`random` module; pass a seeded `random.Random` instance as `rng` for
reproducible runs.
"""

from __future__ import annotations

import random
import time
from typing import Optional

from playwright.sync_api import Page

_DEFAULT_RNG = random.Random()


def _sleep(ms: float) -> None:
    time.sleep(ms / 1000.0)


def _rand(rng: random.Random, mn: float, mx: float) -> float:
    return mn + rng.random() * (mx - mn)


def human_scroll(
    page: Page,
    distance: Optional[float] = None,
    direction: str = "down",
    max_ticks: int = 8,
    rng: Optional[random.Random] = None,
) -> None:
    """
    - distance: total vertical distance in px. Default: one viewport-ish,
      downward.
    - direction: 'down' | 'up'.
    - max_ticks: split the distance into at most this many ticks.
    """
    if rng is None:
        rng = _DEFAULT_RNG
    sign = -1 if direction == "up" else 1
    remaining = abs(distance if distance is not None else _rand(rng, 400, 900))
    ticks = 0

    while remaining > 0 and ticks < max_ticks:
        # Early ticks are larger; the last ones fine-tune, like a real reader.
        delta = (
            _rand(rng, 0.4, 0.6) * remaining
            if ticks == 0
            else _rand(rng, 0.15, 0.45) * remaining
        )
        page.mouse.wheel(0, sign * max(40, delta))
        remaining -= delta
        ticks += 1
        _sleep(_rand(rng, 120, 450))
